<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Methods for developing a machine learning framework for precise 3D domain boundary prediction at base-level resolution</title>
    <meta charset="utf-8" />
    <meta name="author" content="Dissertation Defense  Virginia Commonwealth University | Department of Biostatistics" />
    <link href="libs/remark-css/default.css" rel="stylesheet" />
    <link href="libs/remark-css/default-fonts.css" rel="stylesheet" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# <font color="febe10">Methods for developing a machine learning framework for precise 3D domain boundary prediction at base-level resolution</font>
### <font color="febe10">Dissertation Defense<br> Virginia Commonwealth University | Department of Biostatistics</font><br><img src="img/vcu_mcv_logo.png" width="150" />
### <font color="febe10">Spiro C Stilianoudakis, MA</font><br><svg viewBox="0 0 496 512" xmlns="http://www.w3.org/2000/svg" style="height:1em;fill:currentColor;position:relative;display:inline-block;top:.1em;"> [ comment ] <path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"></path></svg> <a href="https://github.com/stilianoudakis">stilianoudakis</a><br><font color="febe10">February 8, 2021</font>

---




# Outline of Defense

--

1. &lt;small&gt;The 3D genome&lt;/small&gt;

   + &lt;small&gt;Topologically Associated Domains (TADs)&lt;/small&gt;
   + &lt;small&gt;Hi-C sequencing&lt;/small&gt;

--
   
2. &lt;small&gt;Methods for analyzing 3D structure&lt;/small&gt;
   + &lt;small&gt;TAD-callers&lt;/small&gt;
   + &lt;small&gt;Prediction-based methods&lt;/small&gt;

--

3. &lt;small&gt;Developing a machine learning framework for precise TAD boundary prediction at base-level resolution&lt;/small&gt;
   + &lt;small&gt;Aim 1: Develop a machine learning framework to establish an optimal TAD boundary region prediction model&lt;/small&gt;
   + &lt;small&gt;Aim 2: Develop a density-based partitioning technique for precise boundary prediction at base-level resolution&lt;/small&gt;
   + &lt;small&gt;Aim 3: Develop a technique for predicting boundaries on cell lines that do not have publicly available Hi-C data&lt;/small&gt;

--

4. &lt;small&gt;Results&lt;/small&gt;

--

5. &lt;small&gt;Conclusions&lt;/small&gt;

--

6. &lt;small&gt;Deliverables&lt;/small&gt;

--

7. &lt;small&gt;Future work&lt;/small&gt;

---

class: left
background-image: url("img/3d_genome_1.png")
background-size: cover
background-position: center
background-size: contain

# 3D genome

---

class: left
background-image: url("img/3d_genome_2.png")
background-size: cover
background-position: center
background-size: contain

# 3D genome

---

class: left
background-image: url("img/3d_genome_3.png")
background-size: cover
background-position: center
background-size: contain

# 3D genome

---

class: left
background-image: url("img/3d_genome_4.png")
background-size: cover
background-position: center
background-size: contain

# 3D genome

---

# Topologically Associating Domains (TADs)

--

- TADs are stable and dynamic structures 

   + Highly conserved across species and cell lines
   + Associated with gene regulation

--

- Disruption of the boundaries that demarcate TADs has been shown to lead to rare diseases and even cancer

   + A 60kb deletion including a TAD boundary led to overexpression of the LMB1 gene resulting in the rare disorder known as autosomal-dominant adult-onset demyelinating leukodystrophy (ADLD)

--

- Understanding the mechanisms through which TADs are formed remains a top priority in our goal to fully understand the functionality of the human genome

---

# Hi-C Sequencing

.pull-left[
- Hi-C is a technique to capture long-range 3D genomic interactions across the whole genome

- A typical Hi-C experiment produces 20x the number of reads as a standard RNA-seq experiment

- Produces a contact matrix
]

.pull-right[
![](img/contact_matrix.png)
]

---

# Methods for analyzing the 3D structure

--

- 2 primary areas of research:

--

   1. Determining the location of boundaries that demarcate TADs using contact matrices (TAD-callers)

--

   2. Determining the full list of genomic elements that are associated with TAD formation (prediction based methods)

---

class: right
background-image: url("img/tad_caller_1.png")
background-size: cover
background-position: center
background-size: contain

# TAD-callers

---

class: right
background-image: url("img/tad_caller_2.png")
background-size: cover
background-position: center
background-size: contain

# TAD-callers

---

class: right
background-image: url("img/tad_caller_3.png")
background-size: cover
background-position: center
background-size: contain

# TAD-callers

---

# TAD-callers

- Arrowhead (Rao et al.(‎2014))

   + How author explained why finding TAD segmentation from the contact map is tricky:

     &lt;small&gt;"This is due to experimental factors such as noise and inadequate coverage. It is also because of the intrinsic difficulty of the problem: the decline in contact frequency at domain edges can be subtle, and the very rapid decline in contact probability observed as one moves away from the diagonal of a contact map is a major confound for most approaches."&lt;/small&gt;
     
- Additional pitfall:

   + Ignore the fundamental role of functional genomic data (ChIP-seq data) in chromatin folding
   
---

# Prediction-based methods

1. Mourad et al. (PLOS: Computational Biology; 2016)

   + Built on Arrowhead-defined boundaries @ 10 kb
   + Uses only TFBS
   + Percentage of overlap as feature space
   + Multiple logistic regression

2. Hong et al. (Nucleic Acid Research; 2017)

   + Built on Arrowhead-defined boundaries @ 50 kb
   + Uses only TFBS
   + Counts of overlap as feature space
   + `\(l_{2}\)` regularized logistic regression
 
---

# Bridging the gap between 3D HiC data and 1D ChIP-seq data for more precise domain boundary prediction

--

- Unanswered questions:

--

   1. Can a unified framework allow us to more accurately determine the most informative genomic elements of TAD boundary formation? (Aim 1)
      + Feature engineering
      + Resolution 
      + Class imbalance
      
--

   2. Can we leverage the predictive model to more precisely predict TAD boundaries at base-level resolution? (Aim2)

--

   3. Can we extend these methods to predict boundaries on cell lines that do not have Hi-C data available? (Aim 3)

---

# Aim 1: Develop a machine learning framework to establish an optimal TAD boundary region prediction model

- We chose the random forest (RF) machine learning algorithm as our binary classification tool

- Ground truth TAD boundaries were identified using the ARROWHEAD pipeline at 5 kb, 10 kb, 25 kb, 50 kb, and 100 kb for 2 cell lines: GM12878 and K562

- A total of 78 cell line specific functional genomic annotations were used to build the feature space: BroadHMM, HM, and TFBS

---

# Aim 1: Develop a machine learning framework to establish an optimal TAD boundary region prediction model

Defining a resolution-specific response vector **Y** by *shifted binning*

![](img/shifted_binning.png)

---

# Aim 1: Develop a machine learning framework to establish an optimal TAD boundary region prediction model

Defining the feature space `\(\textbf{X} = \{ X_1, X_2, \cdots, X_p \}\)`

&lt;img src="img/feature_engineering_1.png" width="600px" style="display: block; margin: auto;" /&gt;

---

# Aim 1: Develop a machine learning framework to establish an optimal TAD boundary region prediction model

Defining the feature space `\(\textbf{X} = \{ X_1, X_2, \cdots, X_p \}\)`

&lt;img src="img/feature_engineering_2.png" width="600px" style="display: block; margin: auto;" /&gt;

---

# Aim 1: Develop a machine learning framework to establish an optimal TAD boundary region prediction model

.pull-left[
`\(L = \{ GM12878, K562 \}\)`
`\(R = \{ 5 kb, 10 kb, 25 kb, 50 kb, 100 kb \}\)`
`\(P = \{ Signal, OC, OP, Distance \}\)`
`\(S = \{ None, RUS, ROS, SMOTE \}\)`
`\(C = \{ CHR1,CHR2, \cdots \}\)`
]

.pull-right[
![](img/framework.png)
]

---

# Aim 2: Develop a density-based partitioning technique for precise boundary prediction at base-level resolution

- Introducing *preciseTAD*

- What it does…

   + Leverages a predictive model in conjunction with:

      1. Density-based spatial clustering of applications with noise (DBSCAN), and 
      
      2. partitioning around mediods (PAM)
      
    to precisely predict the most biologically meaningful chromosome-specific 	TAD boundaries
    
---

# Aim 2: Develop a density-based partitioning technique for precise boundary prediction at base-level resolution

.pull-left[
How it works...
]

&lt;img src="img/psuedocode.png" width="400px" style="display: block; margin: auto;" /&gt;

---

# Aim 2: Develop a density-based partitioning technique for precise boundary prediction at base-level resolution

- Evaluating precision of boundary prediction

   1. Signal profile plots
   
   2. Enriched heatmaps 
   
   3. Distance to nearest ChIP-seq peak region
   
   4. Venn diagrams/Jaccard Indices
   
---

# Aim 3: Develop a technique for predicting boundaries on cell lines that do not have publicly available Hi-C data

--

- Recall the framework described in Aim 1

--

   + Now we train our RF on data from one cell line and validate it on test data from the other cell line

--

   + We compared 2 separate cases:

--

   1. Models trained on GM12878 and tested on GM12878 vs. models trained on K562 and tested on GM12878, and

--

   2. Models trained on K562 and tested on K562 vs. models trained on GM12878 and tested on K562

--

- Compare performances using AUROC from averaged sensitivities and specificities

- Implement preciseTAD using the same cross-cell line framework

- Compare boundary location for same cell line prediction vs. cross cell line prediction using Jaccard Indices

---

# Results: Aim 1

.pull-left[
Optimal BA is obtained for:

   + (log2) distance-type feature space,
   
   + RUS, and
   
   + 5 kb defined genomic bins
]

.pull-right[
![](img/fig3.png)
]

---

class: left
background-image: url("img/fig4_1.png")
background-size: cover
background-position: center
background-size: 87%

# Results: Aim 1

---

class: left
background-image: url("img/fig4_2.png")
background-size: cover
background-position: center
background-size: 87%

# Results: Aim 1

---

class: left
background-image: url("img/fig4_3.png")
background-size: cover
background-position: center
background-size: 87%

# Results: Aim 1

---

class: left
background-image: url("img/fig5_1.png")
background-size: cover
background-position: center
background-size: 87%

# Results: Aim 2

---

class: left
background-image: url("img/fig5_2.png")
background-size: cover
background-position: center
background-size: 87%

# Results: Aim 2

---

class: left
background-image: url("img/fig6_1.png")
background-size: cover
background-position: center
background-size: 87%

# Results: Aim 2

---

class: left
background-image: url("img/fig6_2.png")
background-size: cover
background-position: center
background-size: 87%

# Results: Aim 2

---

class: left
background-image: url("img/fig6_3.png")
background-size: cover
background-position: center
background-size: 87%

# Results: Aim 2

---

# Results: Aim 2

.pull-left[
- *preciseTAD* is invariant to low resolution ground truth boundaries

- Implementing *preciseTAD* on low resolution Arrowhead results in similar boundaries across resolution
]

.pull-right[
![](img/fig7.png)
]

---

class: left
background-image: url("img/fig8_1.png")
background-size: cover
background-position: center
background-size: 87%

# Results: Aim 3

---

class: left
background-image: url("img/fig8_2.png")
background-size: cover
background-position: center
background-size: 87%

# Results: Aim 3

---

# Conclusions

--

- TAD-calling &amp; boundary prediction is a multi-faceted problem

--

- Conventional TAD-callers focus more on computational efficiency and less on precision

--

- Our framework allows us to determine the optimal list of functional genomic elements most predictive of boundary formation

--

- *preciseTAD* boundaries are:

   + More enriched for top predictive TFBS,

   + More conserved across cell lines,

   + Invariant to low resolution

--

- Our methods can accurately predict TAD boundaries on cell lines that do not have Hi-C available

---

# Deliverables

--

- Models trained using Aim 1 are provided in [preciseTADhub](https://bioconductor.org/packages/devel/data/experiment/html/preciseTADhub.html) as an ExperimentHub R package

--

- Suite of tools discussed are provided in the [preciseTAD](https://bioconductor.org/packages/devel/data/experiment/html/preciseTAD.html) R package

--

- *preciseTAD* was presented as a workshop at the [EuroBioc2020](https://eurobioc2020.bioconductor.org/workshops.html) conference

--

- Our paper has been submitted to Genome Biology

preciseTAD: A machine learning framework for precise 3D domain boundary prediction at base-level resolution
Spiro C. Stilianoudakis, Mikhail G. Dozmorov
bioRxiv 2020.09.03.282186; doi: https://doi.org/10.1101/2020.09.03.282186

---

# Future Work

--

- Incorporate directionality of CTCF motifs

--

- Explore the hierarchy of TADs

--

- Continue to annotate the 3D structure of other cell lines

---

# Acknowledgements 

--

Thank you to my both of my advisors Dr. Mikhail Dozmorov &amp; Dr. Le Kang

--

Thank you to my committee members:

- Dr. Roy Sabo

- Dr. David Wheeler

- Dr. Paul Brooks

- Dr. Tim York

--

My RA Advisors: Dr. Caroline Carrico &amp; Dr. Shilpa Naavaal

--

The friends I made at VCU, notably John Stansfield and Kellen Cresswell

--

Most importantly my family!

---

# Questions?

---

# Appendix

.pull-left[
- Synthetic Minority Over-Sampling Technique

- Incorporates both random under- and over-sampling

- Under-sampling is performed without replacement from the majority class
Over-sampling is performed by creating new synthetic observations using the minority class

- A synthetic example is created by first taking the difference between the minority class feature vector (data point in 2D) under consideration and its nearest neighbor, then multiplying the difference by a random number between 0 and 1, and adding it to the difference
]

.pull-right[
![](img/figA1.png)
]

---

# Appendix

- For each boundary (both called and predicted), a 5000 bp flank was created on either side. 

- The flanking was broken up into 50 bp segments, for a total of 100 windows on both sides of a given boundary. 

- A coverage value was generated for each window with respect to a given ChIP-seq annotation, as the mean signal averaged by the width of the window, denoted by,

`$$v_c = \dfrac{\Sigma_{j}^{m} \Sigma_{i}^{n} x_{ij}w_{ij}}{L}$$`

where `\(x_{ij}\)` is the total number of base pairs of annotation `\(i\)` in window `\(j\)` and `\(w_{ij}\)` is the number of base pairs that overlap between annotation `\(i\)` and window `\(j\)`. The denominator, `\(L\)`, is the width of the windows (here, `\(L\)` = 50). 

- The rows of the heatmaps are ordered by enriched scores. 

- The enriched scores are calculated as the sum of coverage values weighted by the distance to the flanked boundary, denoted by

`$$s_{c} = \Sigma_{d=1}^{n_1} \left( \dfrac{x_{1d} \times d}{n_1} \right) + \Sigma_{u=1}^{n_2} \left( \dfrac{x_{2u} \times (n_2 - u + 1)}{n_2} \right)$$`

---

# Appendix

&lt;img src="img/enriched_heatmaps.png" width="600px" style="display: block; margin: auto;" /&gt;

---

# Appendix

Jaccard Index

- The Jaccard is used as a measure of overlap between two genomic regions A and B defined as

`$$J(A,B) = \dfrac{A \cap B}{A \cup B}$$`

where `\(A \cap B\)` represents the number of base pairs that overlap between `\(A\)` and `\(B\)`, and `\(A \cup B\)` represents the total number of base pairs between `\(A\)` and `\(B\)`

- `\(A\)` and `\(B\)` represent cell line specific flanked boundaries

---

# Evaluation of *preciseTAD* parameters

&lt;img src="img/parameters.png" width="1000px" style="display: block; margin: auto;" /&gt;

---

# References

1. Rao SS, Huntley MH, Durand NC, Stamenova EK, Bochkov ID, Robinson JT, Sanborn AL, Machol I, Omer AD, Lander ES, others: A 3D map of the human genome at kilobase resolution reveals principles of chromatin looping. Cell 2014, 159:1665–1680. 

2. Rosenbloom KR, Sloan CA, Malladi VS, Dreszer TR, Learned K, Kirkup VM, Wong MC, Maddren M, Fang R, Heitner SG, others: ENCODE data in the ucsc genome browser: Year 5 update. Nucleic acids research 2012, 41:D56–D63.

3. Weinreb, C., &amp; Raphael, B. J. (2015). Identification of hierarchical chromatin domains. Bioinformatics, 32(11), 1601-1609.

4. Hansen, A. S., Cattoglio, C., Darzacq, X., &amp; Tjian, R. (2018). Recent evidence that TADs and chromatin loops are dynamic structures. Nucleus, 9(1), 20-32.

5. Durand NC, Shamim MS, Machol I, Rao SS, Huntley MH, Lander ES, Aiden EL: Juicer provides a one-click system for analyzing loop-resolution hi-c experiments. Cell systems 2016, 3:95–98.

6. Rosenbloom KR, Sloan CA, Malladi VS, Dreszer TR, Learned K, Kirkup VM, Wong MC, Maddren M, Fang R, Heitner SG, others: ENCODE data in the ucsc genome browser: Year 5 update. Nucleic acids research 2012, 41:D56–D63.

7. Chawla NV, Bowyer KW, Hall LO, Kegelmeyer WP: SMOTE: Synthetic minority over-sampling technique. Journal of artificial intelligence research 2002, 16:321–357.

8. Lupiáñez DG, Spielmann M, Mundlos S: Breaking tads: How alterations of chromatin domains result in disease. Trends in Genetics 2016, 32:225–237.

9. Dali R, Blanchette M: A critical assessment of topologically associating domain prediction tools. Nucleic acids research 2017, 45:2994–3005.

10. Fraser J, Ferrai C, Chiariello AM, Schueler M, Rito T, Laudanno G, Barbieri M, Moore BL, Kraemer DC, Aitken S, others: Hierarchical folding and reorganization of chromosomes are linked to transcriptional changes in cellular differentiation. Molecular systems biology 2015, 11.

11. Fudenberg G, Imakaev M, Lu C, Goloborodko A, Abdennur N, Mirny LA: Formation of chromosomal domains by loop extrusion. Cell reports 2016, 15:2038–2049.

12. Boulesteix A-L, Janitza S, Kruppa J, König IR: Overview of random forest methodology and practical guidance with emphasis on computational biology and bioinformatics. Wiley Interdisciplinary Reviews: Data Mining and Knowledge Discovery 2012, 2:493–507.

13. Kuhn M: The caret package. R Foundation for Statistical Computing, Vienna, Austria. URL https://cran. r-project. org/package= caret 2012,.

14. Hahsler M, Piekenbrock M, Doran D: Dbscan: Fast density-based clustering with r. Journal of Statistical Software 2019, 25:409–416.

---

# (Un)frequently asked questions

1. Why random forest?

   + &lt;font color="red"&gt;Invariant toward overfitting, doesn't pick out one from group of collinear predictors, interpretable measure of variable importance, flexible hyperparameters&lt;/font&gt;

2. Since random forest is an ensemble of decision trees (nonlinear), why normalize distances?

   + &lt;font color="red"&gt;Normalization was done for computation efficiency&lt;/font&gt;
   
3. What did you set as the other parameters in your random forest models?

   + &lt;font color="red"&gt;ntree=500; nodesize=0.01*nrow(data)&lt;/font&gt;
   
---

# More (un)frequently asked questions

1. Why not perform simulation study instead of relying on "ground truth" boundaries?

   + &lt;font color="red"&gt;While there are methods for simulating Hi-C contact matrices (and therefore, TADs) and methods for simulating ChIP-seq data, they are done through separate processes. Thus, building models using both components will only capture noise.&lt;/font&gt;
   
2. Is it biased to build models using annotations and then used the enrichment of same annotations to gauge performance?

   + &lt;font color="red"&gt;In a vacuum, yes. However, no TAD-caller is evaluated in a vacuum. They are always compared to other TAD-callers. This is because there does not exist a unified metric of performance.&lt;/font&gt;

3. How can you be sure that you are not predicting every boundary with a CTCF, RAD21, etc. overlap?

   + &lt;font color="red"&gt;There are hundreds of thousands of ChIP-seq peak regions throughout the genome. That would result in hundreds of thousands of predicted boundaries. We see from figures like on Slide 29 that this is not the case. Concepts like the probability threshold (t) and normalized (NE) are used to guard against this.&lt;/font&gt;
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"ratio": "16:9",
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
